### DISCUSSION TITLE
Can't update an image (401 Unauthorized)

### DISCUSSION THREAD

**Discussion by rohitp30** [containrrr/watchtower#1097] [Q&A]
Hey, so I set up Watchtower and I keep getting this 404 error when trying to pull down the image. The Docker Project is public and I tried the command without supplying my credentials and I still got the same error. I already did `docker login` and the project exists on Docker Hub. Can't really seem to find a solution to this.

<img width="471" alt="image" src="https://user-images.githubusercontent.com/56415283/137650419-9bfe27c1-b901-4e6e-baea-f79fe98d5580.png">

Command:
```bash
docker run -d \
  --name watchtower \
  -e REPO_USER=spaceturtle0 \
  -e REPO_PASS=... \
  -v /var/run/docker.sock:/var/run/docker.sock \
  -e WATCHTOWER_NOTIFICATIONS=shoutrrr \
  -e WATCHTOWER_NOTIFICATION_URL="..." \
  -e WATCHTOWER_NOTIFICATION_TEMPLATE="{{range .}}{{.Time.Format \"2006-01-02 15:04:05\"}} ({{.Level}}): {{.Message}}{{println}}{{end}}" \
  -e WATCHTOWER_POLL_INTERVAL=60 \
  containrrr/watchtower ss-timmy --debug
```

Logs:
https://gist.github.com/Space-Turtle0/69c07959fd311cb5db9c85aedb11e632

Is there something I'm missing here?

**Comment by piksel** [comment]
https://hub.docker.com/_/timmy doesn't seem to exist? What is the image name you are trying to use?

**Reply by rohitp30** [reply]
spaceturtle0/ss-timmy

Although when I tried specifying that in the run command, Watchtower didn't update anything. Looks like it didn't find an image with that name or something.

**Reply by piksel** [reply]
No, it's in your original container run command, i.e. something like `docker run spaceturtle0/ss-timmy:latest`. You have probably started it with a locally tagged `timmy:latest` which watchtower cannot pull.

**Reply by rohitp30** [reply]
Same issue:
https://gist.github.com/Space-Turtle0/e894a2bcb9e6709be85d54b4e6338f82

This time ran it with the image name `spaceturtle0/ss-timmy`

**Reply by piksel** [reply]
You are passing the image tag as container name (`spaceturtle0/ss-timmy`) to watchtower:

> Using notifications: discord
> Only checking containers with name "spaceturtle0/ss-timmy"

Run your container with:
```
docker run --name <CONTAINER-NAME> <IMAGE-TAG>
```
i.e.
```
docker run --name timmy spaceturtle0/ss-timmy:latest
```

then run watchtower with the container name (or skip it to check all):
```
docker run -d -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower <CONTAINER-NAME>
```
i.e.
```
docker run -d -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower timmy
```

**Reply by rohitp30** [reply]
I used `docker run -d -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower timmy`

Logs:
```
time="2021-10-18T17:10:13Z" level=info msg="Watchtower 1.3.0\nUsing notifications: discord\nOnly checking containers with name \"timmy\"\nScheduling first run: 2021-10-18 17:10:23 +0000 UTC\nNote that the first check will be performed in 9 seconds"
time="2021-10-18T17:10:23Z" level=debug msg="Checking containers for updated images"
time="2021-10-18T17:10:23Z" level=debug msg="Retrieving running containers"
time="2021-10-18T17:10:23Z" level=debug msg="Session done: 0 scanned, 0 updated, 0 failed"
time="2021-10-18T17:10:23Z" level=debug msg="Scheduled next run: 2021-10-18 17:10:33 +0000 UTC"
time="2021-10-18T17:10:33Z" level=debug msg="Checking containers for updated images"
```

**Reply by rohitp30** [reply]
`docker ps` output:
```
CONTAINER ID   IMAGE                   COMMAND                  CREATED         STATUS         PORTS                                       NAMES
05b151207232   containrrr/watchtower   "/watchtower timmy -…"   3 minutes ago   Up 3 minutes   8080/tcp                                    watchtower
4245340db460   timmy                   "python3 main.py"        3 days ago      Up 3 days                                                  ss-timmy
9a82e7542584   registry:2              "/entrypoint.sh /etc…"   3 days ago      Up 3 days      0.0.0.0:5000->5000/tcp, :::5000->5000/tcp   registry
74b92affaea3   bannerbot               "docker-entrypoint.s…"   4 days ago      Up 4 days                                                  brave_einstein
```

**Reply by rohitp30** [reply]
I'll try running it without the name checks, looks like it can't find any matching images.

**Reply by piksel** [reply]
Exactly...

```yaml
CONTAINER-ID: 4245340db460
IMAGE: timmy
COMMAND: "python3 main.py"
CREATED: 3 days ago
STATUS: Up 3 days
PORTS:
NAMES: ss-timmy
```
You started your container with the **image** tag `timmy` and not `spaceturtle0/ss-timmy:latest`.
The container **name(s)** are `ss-timmy` (which is what watchtower takes as an argument).

**Reply by piksel** [reply]
Locally you can tag the images however you like, but dockerhub does not know what image `timmy` is.

**Reply by rohitp30** [reply]
Alright, I'll try it out.

**Reply by rohitp30** [reply]
So just to confirm:
`docker run --name timmy spaceturtle0/ss-timmy:latest`

then:
`docker run -d -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower ss-timmy`?

**Reply by piksel** [reply]
Almost! The argument to `--name` in the first should match the last argument on the second, so set both to either `ss-timmy` or `timmy`

**Reply by rohitp30** [reply]
`docker run -d -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower timmy` should be correct then, right?

**Reply by rohitp30** [reply]
```
root@schoolsimplifiedbots:~# docker logs /watchtowertime="2021-10-18T17:49:28Z" level=debugtime="2021-10-18T17:49:28Z" level=debug msg="Sleeping for a second to ensure the docker api client has been properly initialized."time="2021-10-18T17:49:29Z" level=debug msg="Making sure everything is sane before starting"time="2021-10-18T17:49:29Z" level=debug msg="Retrieving running containers"
time="2021-10-18T17:49:29Z" level=debug msg="There are no additional watchtower containers"
time="2021-10-18T17:49:29Z" level=debug msg="Watchtower HTTP API skipped."
time="2021-10-18T17:49:29Z" level=info msg="Watchtower 1.3.0\nUsing notifications: discord\nOnly checking containers with name \"timmy\"\nScheduling first run: 2021-10-18 17:49:39 +0000 UTC\nNote that the first check will be performed in 9 seconds"
time="2021-10-18T17:49:39Z" level=debug msg="Checking containers for updated images"
time="2021-10-18T17:49:39Z" level=debug msg="Retrieving running containers"
time="2021-10-18T17:49:39Z" level=debug msg="Session done: 0 scanned, 0 updated, 0 failed"
time="2021-10-18T17:49:39Z" level=debug msg="Scheduled next run: 2021-10-18 17:49:49 +0000 UTC"
time="2021-10-18T17:49:49Z" level=debug msg="Checking containers for updated images"
time="2021-10-18T17:49:49Z" level=debug msg="Retrieving running containers"
time="2021-10-18T17:49:49Z" level=debug msg="Session done: 0 scanned, 0 updated, 0 failed"
time="2021-10-18T17:49:49Z" level=debug msg="Scheduled next run: 2021-10-18 17:49:59 +0000 UTC"
```

Didn't seem to work with this command:
```bash
docker run -d \
  --name watchtower \
  -v /var/run/docker.sock:/var/run/docker.sock \
  -e WATCHTOWER_NOTIFICATIONS=shoutrrr \
  -e WATCHTOWER_NOTIFICATION_URL="..." \
  -e WATCHTOWER_NOTIFICATION_TEMPLATE="{{range .}}{{.Time.Format \"2006-01-02 15:04:05\"}} ({{.Level}}): {{.Message}}{{println}}{{end}}" \
  -e WATCHTOWER_POLL_INTERVAL=10 \
  containrrr/watchtower timmy --debug
```

AAAHHH

**Reply by rohitp30** [reply]
I mean even with the naming issue, it still can't pull the image without any name checks.

**Reply by piksel** [reply]
I tried it locally. 🤷‍♀️

<details>
<summary><code>$ docker run -d --name timmy spaceturtle0/ss-timmy:latest</code></summary>

```
Unable to find image 'spaceturtle0/ss-timmy:latest' locally
latest: Pulling from spaceturtle0/ss-timmy
1c47a4233665: Pull complete
580c36351f83: Pull complete
1b655be75802: Pull complete
831021d515bc: Pull complete
4cc5f3294b89: Pull complete
cafbfc2a7baf: Pull complete
ee8127f702af: Pull complete
c6dce4367c41: Pull complete
99dd2e2c6642: Pull complete
0aa3f182b621: Pull complete
6b72e97c6bb3: Pull complete
5a59e5813c79: Pull complete
a928c5bb6066: Pull complete
733bc768608a: Pull complete
Digest: sha256:29f2e88efec969f1d087f6396aa32c1e3166eea3016a9d62769a20595175edb8
Status: Downloaded newer image for spaceturtle0/ss-timmy:latest
WARNING: The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64) and no specific platform was requested
98601ae074a07ee41052443275c01c0ac35755c5107aa43073a25465af192129
```

</details>

<details>
<summary><code>$ docker run -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower --debug --run-once timmy</code></summary>

```
time="2021-10-18T18:02:44Z" level=debug
time="2021-10-18T18:02:44Z" level=debug msg="Sleeping for a second to ensure the docker api client has been properly initialized."
time="2021-10-18T18:02:45Z" level=debug msg="Making sure everything is sane before starting"
time="2021-10-18T18:02:45Z" level=info msg="Watchtower 1.3.0\nUsing no notifications\nOnly checking containers with name \"timmy\"\nRunning a one time update."
time="2021-10-18T18:02:45Z" level=debug msg="Checking containers for updated images"
time="2021-10-18T18:02:45Z" level=debug msg="Retrieving running containers"
time="2021-10-18T18:02:45Z" level=debug msg="Trying to load authentication credentials." container=/timmy image="spaceturtle0/ss-timmy:latest"
time="2021-10-18T18:02:45Z" level=debug msg="No credentials for spaceturtle0 found" config_file=/config.json
time="2021-10-18T18:02:45Z" level=debug msg="Got image name: spaceturtle0/ss-timmy:latest"
time="2021-10-18T18:02:45Z" level=debug msg="Checking if pull is needed" container=/timmy image="spaceturtle0/ss-timmy:latest"
time="2021-10-18T18:02:45Z" level=debug msg="Building challenge URL" URL="https://index.docker.io/v2/"
time="2021-10-18T18:02:45Z" level=debug msg="Got response to challenge request" header="Bearer realm=\"https://auth.docker.io/token\",service=\"registry.docker.io\"" status="401 Unauthorized"
time="2021-10-18T18:02:45Z" level=debug msg="Checking challenge header content" realm="https://auth.docker.io/token" service=registry.docker.io
time="2021-10-18T18:02:45Z" level=debug msg="Setting scope for auth token" image=spaceturtle0/ss-timmy scope="repository:spaceturtle0/ss-timmy:pull"
time="2021-10-18T18:02:45Z" level=debug msg="No credentials found."
time="2021-10-18T18:02:46Z" level=debug msg="Parsing image ref" host=index.docker.io image=spaceturtle0/ss-timmy normalized="docker.io/spaceturtle0/ss-timmy:latest" tag=latest
time="2021-10-18T18:02:46Z" level=debug msg="Doing a HEAD request to fetch a digest" url="https://index.docker.io/v2/spaceturtle0/ss-timmy/manifests/latest"
time="2021-10-18T18:02:46Z" level=debug msg="Found a remote digest to compare with" remote="sha256:29f2e88efec969f1d087f6396aa32c1e3166eea3016a9d62769a20595175edb8"
time="2021-10-18T18:02:46Z" level=debug msg=Comparing local="sha256:29f2e88efec969f1d087f6396aa32c1e3166eea3016a9d62769a20595175edb8" remote="sha256:29f2e88efec969f1d087f6396aa32c1e3166eea3016a9d62769a20595175edb8"
time="2021-10-18T18:02:46Z" level=debug msg="Found a match"
time="2021-10-18T18:02:46Z" level=debug msg="No pull needed. Skipping image."
time="2021-10-18T18:02:46Z" level=debug msg="No new images found for /timmy"
time="2021-10-18T18:02:46Z" level=debug msg="Session done: 1 scanned, 0 updated, 0 failed"
```

</details>

**Reply by piksel** [reply]
> I mean even with the naming issue, it still can't pull the image without any name checks.

Have you stopped your old `timmy` container then? That one can never be updated and will will always trigger the initial error you posted. When running `docker ps` there should only be a "timmy"-container with
```yaml
IMAGE: spaceturtle0/ss-timmy:latest
NAMES: timmy
```

**Reply by rohitp30** [reply]
```
root@schoolsimplifiedbots:~# docker run -d --name timmy spaceturtle0/ss-timmy:latest
WARNING: The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64) and no specific platform was requested
f7ef6cc011afc1ab74f2b51afede7e060bfcd766436e21954a0c5a3704697e2a

root@schoolsimplifiedbots:~# docker run -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower --debug --run-once timmy
time="2021-10-18T18:13:56Z" level=debug
time="2021-10-18T18:13:56Z" level=debug msg="Sleeping for a second to ensure the docker api client has been properly initialized."
time="2021-10-18T18:13:57Z" level=debug msg="Making sure everything is sane before starting"
time="2021-10-18T18:13:57Z" level=info msg="Watchtower 1.3.0\nUsing no notifications\nOnly checking containers with name \"timmy\"\nRunning a one time update."
time="2021-10-18T18:13:57Z" level=debug msg="Checking containers for updated images"
time="2021-10-18T18:13:57Z" level=debug msg="Retrieving running containers"
time="2021-10-18T18:13:57Z" level=debug msg="Session done: 0 scanned, 0 updated, 0 failed"
```

**Reply by rohitp30** [reply]
> Have you stopped your old `timmy` container then? That one can never be updated and will will always trigger the initial error you posted. When running `docker ps` there should only be a "timmy"-container with
> 
> ```yaml
> IMAGE: spaceturtle0/ss-timmy:latest
> NAMES: timmy
> ```

Yeah I've stopped it.

**Reply by piksel** [reply]
What does `docker ps` say? And what happens if you omit `timmy` from the watchtower command? You can run it with `--monitor-only` to avoid having it update your other containers:
```
docker run -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower --debug --run-once --monitor-only
```

**Reply by rohitp30** [reply]
Docker PS:
```
CONTAINER ID   IMAGE        COMMAND                  CREATED             STATUS             PORTS                                       NAMES
3c8763a41286   timmy        "python3 main.py"        About an hour ago   Up About an hour                                               magical_chaum
9a82e7542584   registry:2   "/entrypoint.sh /etc…"   3 days ago          Up 3 days          0.0.0.0:5000->5000/tcp, :::5000->5000/tcp   registry
74b92affaea3   bannerbot    "docker-entrypoint.s…"   4 days ago          Up 4 days                                                      brave_einstein
```

CMD Output:
https://gist.github.com/Space-Turtle0/09e5b105f890acb63e4f0f3019d50e29

**Reply by piksel** [reply]
Yeah, I don't know how you are doing it but the image tag is still `timmy` and the name is randomly generated (`magical_chaum`)

```yaml
CONTAINER-ID: 3c8763a41286
IMAGE: timmy
COMMAND: "python3 main.py"
CREATED: About an hour ago
STATUS: Up About an hour
PORTS: 
NAMES: magical_chaum
```

**Reply by rohitp30** [reply]
What if we try without tagging the image, is there something to untag and just run it with its original name (ss-timmy)?

**Reply by piksel** [reply]
What do you mean by "tagging the image"? You should run the container with the dockerhub image tag: `spaceturtle0/ss-timmy:latest`.
The container ID returned from starting with `docker run -d --name timmy spaceturtle0/ss-timmy:latest` (in your log above) does not match the one in the output from `docker ps`. 

Are you running some other commands as well?

**Reply by rohitp30** [reply]
Like doing:
`docker run -d spaceturtle0/ss-timmy:latest`

and:
`docker run -d -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower spaceturtle0/ss-timmy:latest`

**Reply by piksel** [reply]
The argument to watchtower is not an image tag. No container will ever match `spaceturtle0/ss-timmy:latest` since it's an invalid container name.

**Reply by rohitp30** [reply]
Is this output expected when there isn't an available update?
https://gist.github.com/Space-Turtle0/09e5b105f890acb63e4f0f3019d50e29

**Reply by piksel** [reply]
No.

> time="2021-10-18T18:57:52Z" level=debug msg="Pulling image" container=/magical_chaum image="timmy:latest"

It's still running with the image tag `timmy:latest` instead of `spaceturtle0/ss-timmy:latest` and hence cannot be updated, since there is no `timmy` image on dockerhub.

**Reply by rohitp30** [reply]
Alright, I might have found a solution:

The main problem was that I docker kept telling me that I couldn't run the image due the build being on ARM and it can't be built on Ryzen chips (which is present on the server)

Followed these steps to resolve that:
(Basically I didn't pull initally) 
- CD into your local copy of the image (cd Timmy-SchoolSimplified)

- Build the image (docker build -t spaceturtle0/ss-timmy .)

- Remove any pre-existing images if found and then run (docker run -d --name timmy spaceturtle0/ss-timmy:latest)

- docker ps to check results


Results:
```
CONTAINER ID   IMAGE                          COMMAND                  CREATED         STATUS         PORTS                                       NAMES
913a1c081906   spaceturtle0/ss-timmy:latest   "python3 main.py"        3 seconds ago   Up 3 seconds                                               timmy
```

And the `docker run -d -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower spaceturtle0/ss-timmy:latest` found the container this time!

**Reply by rohitp30** [reply]
Seems to work?

<img width="598" alt="image" src="https://user-images.githubusercontent.com/56415283/137825602-28bd9805-12e1-40a7-be5e-c23093ed2f49.png">

**Reply by rohitp30** [reply]
I guess the final question is, is there a way to auto-restart the image after it's been updated? The docker image is a discord bot and it doesn't seem to turn back on regardless of the environment option being set to true: 

My CMD:
```shell
docker run -d \
  --name watchtower \
  -v /var/run/docker.sock:/var/run/docker.sock \
  -e WATCHTOWER_NOTIFICATIONS=shoutrrr \
  -e WATCHTOWER_NOTIFICATION_URL="." \
  -e WATCHTOWER_NOTIFICATION_TEMPLATE="{{range .}}{{.Time.Format \"2006-01-02 15:04:05\"}} ({{.Level}}): {{.Message}}{{println}}{{end}}" \
  -e WATCHTOWER_POLL_INTERVAL=60 \
  -e WATCHTOWER_INCLUDE_RESTARTING=true \
  containrrr/watchtower timmy --debug
```

**Reply by piksel** [reply]
It should always restart, the `WATCHTOWER_INCLUDE_RESTARTING` is just whether it should try to update containers that is currently in the middle of restarting.  Perhaps there is something in the logs that tell you what happens after "Creating /timmy"?

**Reply by piksel** [reply]
Wait... if you have to build it locally to run it, then the image you pull will have the same problem... 
You should probably specify `linux/amd64` as the platform when building (preferably using `buildx`: https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/). If you still need to run on arm you can supply both using `--platform linux/amd64,linux/arm64`

**Reply by rohitp30** [reply]
> It should always restart, the `WATCHTOWER_INCLUDE_RESTARTING` is just whether it should try to update containers that is currently in the middle of restarting. Perhaps there is something in the logs that tell you what happens after "Creating /timmy"?

These are my logs: https://gist.github.com/Space-Turtle0/1218dab5d180ff48beaddea4ea396f04



> Wait... if you have to build it locally to run it, then the image you pull will have the same problem...
> You should probably specify `linux/amd64` as the platform when building (preferably using `buildx`: https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/). If you still need to run on arm you can supply both using `--platform linux/amd64,linux/arm64`

This is passed in via the build command right?
